address_book/internal/service/service.go

```diff
@@ -0,0 +1,125 @@
+package service
+
+import (
+	"addressbook/internal/pb"
+	"context"
+	"fmt"
+	"regexp"
+	"sync"
+)
+
+type AddressBookService struct {
+	pb.AddressBookServiceServer
+	data sync.Map
+}
+
+func (a *AddressBookService) AddContact(_ context.Context, in *pb.AddContactRequest) (*pb.AddContactResponse, error) {
+	if in == nil || in.Contact == nil {
+		return &pb.AddContactResponse{
+			Msg: "Contact has not been added",
+		}, nil
+	}
+
+	a.data.Store(in.Contact.Phone, in.Contact)
+
+	return &pb.AddContactResponse{
+		Msg: "Contact added successfully",
+	}, nil
+}
+
+func (a *AddressBookService) FindContact(_ context.Context, in *pb.FindContactRequest) (*pb.FindContactResponse, error) {
+	if in == nil || in.Query == "" {
+		return &pb.FindContactResponse{
+			Msg: "Empty query, contact has not been found",
+		}, nil
+	}
+
+	contacts := []*pb.Contact{}
+
+	switch in.SearchType {
+	case pb.FindContactRequest_NAME:
+		a.data.Range(func(key interface{}, value interface{}) bool {
+			if contact, ok := value.(*pb.Contact); ok {
+				if contact.Name == in.Query {
+					contacts = append(contacts, contact)
+				}
+			}
+			return true
+		})
+	case pb.FindContactRequest_PHONE:
+		matchString := in.Query
+
+		a.data.Range(func(key interface{}, value interface{}) bool {
+			if contact, ok := value.(*pb.Contact); ok {
+				matched, err := regexp.MatchString(matchString, contact.Phone)
```

Comment:
Wildcards are not the same as regexp.

Reply:
Done

Reply:
Does this fix properly handle * wildcard?

Reply:
you can use ? and * symbols for searching